<a href="https://colab.research.google.com/github/AshmithaMB/Stock-Analysis/blob/main/Micrsoft_Stock_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Course Project on EDA with Python
### Analysis on Microsoft Stocks

Importing Pandas and Plotly 
Through the below command

In [2]:
import pandas as pd

In [3]:
import plotly.graph_objects as go
import plotly.express as px

*Yahoo Finance is one of the most popular websites to collect stock price data* 

*yfinance is an API provided by Yahoo Finance to collect the latest stock price data*

*To use this API, I installed it by using the pip command*

In [4]:
pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 50 kB 2.6 MB/s 
     |████████████████████████████████| 4.0 MB 8.2 MB/s 
     |████████████████████████████████| 112 kB 61.4 MB/s 
     |████████████████████████████████| 128 kB 82.9 MB/s 
     |████████████████████████████████| 110 kB 75.4 MB/s 
     |████████████████████████████████| 62 kB 1.3 MB/s 
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Successfully uninstalled requests-2.23.0
  Attempting uninstall: html5lib
    Found existing installation: html5lib 1.0.1
    Uninstalling html5lib-1.0.1:
      Successfully uninstalled html5lib-1.0.1
  Attempting uninstall: beautifulsoup4
    Found existing installation: beautifulsoup4 4.6.3
    Uninstalling beautifulsoup4-4.6.3:
      Successfully uninstalled beautifulsoup4-4.6.3


In [5]:
import yfinance as yf

*In Python, Date and Time are not data types of their own, but a module named DateTime can be imported to work with the date as well as time. Datetime module comes built into Python, so there is no need to install it externally. The DateTime module provides some functions to get the current date as well as time.*

In [6]:
import datetime
from datetime import date, timedelta

*I have imported date class from the datetime module because the object of datetime class can access strftime() method*

**timedelta( )**  *function under datetime library is used for calculating differences in dates and also can be used for date manipulations.* 

**strftime( )**
*Converts a datetime object containing current date and time to different string formats; %Y, %m, %d etc are format codes. The strftime() method takes one or more format codes as an argument and returns a formatted string based on it.*

**today( )**
*method of date class under DateTime module returns a date object which contains the value of Today’s date*

In [7]:
today = date.today()
#The date object containing current date and time is stored in 'today' variable

d1 = today.strftime("%Y-%m-%d") 
# %Y	Year with century as a decimal number | %-m	Month as a decimal number | %-d	Day of the month as a decimal number
end_date = d1

d2 = date.today() - timedelta(days=365)
d2 = d2.strftime("%Y-%m-%d")
start_date = d2

In [14]:
data = yf.download('MSFT', 
                      start=start_date, 
                      end=end_date, 
                      progress=False)

data["Date"] = data.index

data = data[["Date", "Open", "High", "Low", 
             "Close", "Adj Close", "Volume"]]

data.reset_index(drop=True, inplace=True)

data.head()

Date        Open        High         Low       Close   Adj Close  \
0 2021-12-20  320.049988  322.799988  317.570007  319.910004  316.940247   
1 2021-12-21  323.290009  327.730011  319.799988  327.290009  324.251709   
2 2021-12-22  328.299988  333.609985  325.750000  333.200012  330.106873   
3 2021-12-23  332.750000  336.390015  332.730011  334.690002  331.583008   
4 2021-12-27  335.459991  342.480011  335.429993  342.450012  339.270966   

     Volume  
0  28326500  
1  24740600  
2  24831500  
3  19617800  
4  19947000

In [15]:
data.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

In [22]:
data.describe()

Open        High         Low       Close   Adj Close  \
count  252.000000  252.000000  252.000000  252.000000  252.000000   
mean   272.389563  275.794484  268.591349  272.218651  270.738314   
std     28.073917   27.948155   27.717539   27.950550   27.173756   
min    217.550003  220.410004  213.429993  214.250000  213.647903   
25%    248.490002  252.369995  246.045002  249.949997  249.719185   
50%    268.345001  273.065002  265.115005  268.744995  267.422119   
75%    293.109993  295.622490  289.602501  294.440010  292.260513   
max    343.149994  344.299988  340.320007  342.450012  339.270966   

             Volume  
count  2.520000e+02  
mean   3.112526e+07  
std    1.153723e+07  
min    9.200800e+06  
25%    2.323672e+07  
50%    2.903680e+07  
75%    3.524855e+07  
max    9.042890e+07

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252 entries, 0 to 251
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       252 non-null    datetime64[ns]
 1   Open       252 non-null    float64       
 2   High       252 non-null    float64       
 3   Low        252 non-null    float64       
 4   Close      252 non-null    float64       
 5   Adj Close  252 non-null    float64       
 6   Volume     252 non-null    int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 13.9 KB


In [27]:
data.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

In [17]:
data.shape

(252, 7)

In [21]:
data.Date

0     2021-12-20
1     2021-12-21
2     2021-12-22
3     2021-12-23
4     2021-12-27
         ...    
247   2022-12-13
248   2022-12-14
249   2022-12-15
250   2022-12-16
251   2022-12-19
Name: Date, Length: 252, dtype: datetime64[ns]

From the above Dataset,
* The table shows open, high, low, and closing prices for each period
* The dataset does not have any null values
* Timeframe - December 20th 2021 to December 19th 2022

Attributes:
* Date - Date
* Open - Opening price in the given timeframe
* High - Highest price attained in the given timeframe 
* Low - lowest price attained in the given timeframe 
* Close - Closing price in the given timeframe
* Adj-Close - Stock price after paying off the dividends
* Volume - trading volume for the selected period

Plotting through a Candlestick Chart as it is widely used to analyse the price movements of stock prices, currencies, and other financial instruments

In [9]:
figure = go.Figure(data=[go.Candlestick(x=data["Date"],
                                        open=data["Open"], high=data["High"],
                                        low=data["Low"], close=data["Close"])])

figure.update_layout(title = "Microsoft Stock Price Analysis", xaxis_rangeslider_visible=False)

figure.show()

Plotting through a Bar Plot to analyse continuous variables in the dataframe (open, high, low, close, adj-close), discrete variables(volume)

In [28]:
figure = px.bar(data, x = "Date", y= "Close")

figure.show()

Line Plot using RangeSlider allows us to check the price variation date-wise


As we can see here:

* The closing rate fell drastically from 342.2 in December 2021 to  240.45 in December 2022

* The lowest closing rate was 214.25 observed in November 3rd 2022.
* While the highest was in December 27th 2021, 342.2

In [29]:
figure = px.line(data, x='Date', y='Close', 
                 title='Microsoft Stock Market Analysis with Rangeslider')

figure.update_xaxes(rangeslider_visible=True)

figure.show()

Through a Line Plot here, relationship between two variables in consideration (Date, Open) Date and the opening price of the stock can be analyzed.

In [31]:
figure = px.line(data, x='Date', y='Open', 
                 title='Stock Market Analysis with Time Period Selectors')

figure.update_xaxes(
    rangeselector=dict(
        buttons=list([
            dict(count=1, label="1m", step="month", stepmode="backward"),
            dict(count=6, label="6m", step="month", stepmode="backward"),
            dict(count=3, label="3m", step="month", stepmode="backward"),
            dict(count=1, label="1y", step="year", stepmode="backward"),
            dict(step="all")
            ])
        )
    )
figure.show()

Observations:
* 343.15 was the highest opening price which fell down to as low as 217.55

Using Scatter Plot here to additionally find the outliers

In [36]:
figure = px.scatter(data, x='Date', y='Close', range_x=['2021-07-12', '2022-07-11'],
                 title="Stock Market Analysis by Hiding Weekend Gaps")

figure.update_xaxes(
    rangebreaks=[
        dict(bounds=["sat", "sun"])]
        )

figure.show()

Observations
* There is a downward trend/ pattern in the above graph which could indicate weak to moderate negative correlation between the date and the closing rates
* By hiding the weekend gaps, it is observable that the closing rates at the end of the year are high in comparision to the start of the year